### PR TITLE
Fetched News Data From News APi-Us and saved into Mongo db.

### PR DESCRIPTION
**Pull request by Arundhathy-k**
Fetched News Data From News APi-Us and saved into Mongo db successfully.

### CODE REVIEW

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                .id(newsDto.getId())
+                .status(newsDto.getStatus())
+                .totalResults(newsDto.getTotalResults())
+                .publishedAt(newsDto.getPublishedAt() != null ? newsDto.getPublishedAt() : null)
+                .articles(mapArticlesToEntity(newsDto.getArticles()))
+                .build();
+    }
+
+    private List<NewsDto.Article> mapArticlesToDto(List<NewsEntity.Article> articles) {
+        return articles != null ? articles.stream()
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: use optional to check null

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                        .urlToImage(article.getUrlToImage())
+                        .publishedAt(article.getPublishedAt())
+                        .content(article.getContent())
+                        .build())
+                .collect(Collectors.toList()) : null;
+    }
+
+
+    private List<NewsEntity.Article> mapArticlesToEntity(List<NewsDto.Article> articles) {
+        return articles != null ? articles.stream()
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: use optional to check null

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                        .urlToImage(article.getUrlToImage())
+                        .publishedAt(article.getPublishedAt())
+                        .content(article.getContent())
+                        .build())
+                .collect(Collectors.toList()) : null;
+    }
+
+
+    private NewsDto.Article.Source mapSourceToDto(NewsEntity.Article.Source source) {
+        return source != null ? NewsDto.Article.Source.builder()
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: use optional to check null

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                .id(source.getId())
+                .name(source.getName())
+                .build() : null;
+    }
+
+    private NewsEntity.Article.Source mapSourceToEntity(NewsDto.Article.Source source) {
+        return source != null ? NewsEntity.Article.Source.builder()
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: use optional to check null

File: newsService/src/main/java/com/kovan/app/service/NewsService.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+
+@Service
+public class NewsService {
+
+    private final RestTemplate restTemplate = new RestTemplate();
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Autowired
+    private NewsRepositoryService service;
```
Comment (dhanyesh) [inline newsService/src/main/java/com/kovan/app/service/NewsService.java]: use constructor injection

File: newsService/src/main/java/com/kovan/app/service/NewsService.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        do {
+            pageSize += 20;
+            String apiUrl = buildUrl(page++);
+            String response = restTemplate.getForObject(apiUrl, String.class);
+            try {
+
+                newsDto = objectMapper.readValue(response, NewsDto.class);
+                articles.addAll(newsDto.getArticles());
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException(e);
```
Comment (dhanyesh) [inline newsService/src/main/java/com/kovan/app/service/NewsService.java]: write custom exception class

File: newsService/src/main/java/com/kovan/app/service/NewsService.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        }
+            return  newsDto;
+    }
+    private boolean isNewsAlreadyInDb(String publishedAt) {
+        return newsRepository.findByPublishedAt(publishedAt).isPresent();
+    }
+
+    private String buildUrl(int page) {
+        return "https://newsapi.org/v2/top-headlines?country=us&category=technology&" +
```
Comment (dhanyesh) [inline newsService/src/main/java/com/kovan/app/service/NewsService.java]: find any alternate way to do this and keep url and api key in properties file.

File: mongo-dao/src/main/java/com/kovan/entity/NewsEntity.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    private String id;
+
+    private String status;
+
+    private int totalResults;
+
+    private String publishedAt;
+
+    private List<Article> articles;
+
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/entity/NewsEntity.java:30]: add Instant createdDate , createdBy , Instant UpdatedDate, UpdatedBy

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    public NewsDto toDto(NewsEntity newsEntity) {
+        if (Objects.isNull(newsEntity)) {
+            return null;
+        }
+        return NewsDto.builder()
+                .id(newsEntity.getId())
+                .status(newsEntity.getStatus())
+                .totalResults(newsEntity.getTotalResults())
+                .publishedAt(newsEntity.getPublishedAt() != null ? newsEntity.getPublishedAt() : null)
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: check null using StringUtils

File: mongo-dao/src/main/java/com/kovan/service/NewsRepositoryService.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Autowired
+    private NewsRepository newsRepository;
+
+    @Autowired
+    private NewsMapper newsMapper;
+
+    public NewsDto saveNewsInDb(NewsDto newsDto) {
+        if (Objects.isNull(newsDto)) {
+            throw new IllegalArgumentException("NewsDto cannot be null");
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/service/NewsRepositoryService.java]: check using custom exception

**Review by dhanyesh — CHANGES_REQUESTED** [review]

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java (hunk 2 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: Optional used to check null.

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java (hunk 3 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: Optional used to check null.

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java (hunk 4 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: Optional used to check null.

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java (hunk 9 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: used StringUtils to null check

File: mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java (hunk 1 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/mapper/NewsMapper.java]: Optional used to check null.

File: newsService/src/main/java/com/kovan/app/service/NewsService.java (hunk 6 above)
Comment (Arundhathy-k) [reply newsService/src/main/java/com/kovan/app/service/NewsService.java]: Written custom exception class.

File: mongo-dao/src/main/java/com/kovan/service/NewsRepositoryService.java (hunk 10 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/service/NewsRepositoryService.java]: Used custom exception for exception handling.

File: mongo-dao/src/main/java/com/kovan/entity/NewsEntity.java (hunk 8 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/entity/NewsEntity.java:30]: Added  Instant createdDate , createdBy , Instant UpdatedDate, UpdatedBy.

File: newsService/src/main/java/com/kovan/app/service/NewsService.java (hunk 5 above)
Comment (Arundhathy-k) [reply newsService/src/main/java/com/kovan/app/service/NewsService.java]: Used constructor injection.

File: mongo-dao/src/main/java/com/kovan/dto/NewsDto.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class NewsDto {
+
+    @Id
+    private String id;
```
Comment (dhanyesh) [inline mongo-dao/src/main/java/com/kovan/dto/NewsDto.java:21]: why do we need this Id?

File: newsService/src/main/java/com/kovan/app/service/NewsService.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+        } while (pageSize <= newsDto.getTotalResults());
+
+        LocalDate dateOnly = LocalDate.parse(
+                articles.get(0).getPublishedAt(),
+                DateTimeFormatter.ISO_DATE_TIME
+        );
+
+        if (!isNewsAlreadyInDb(dateOnly.toString())) {
+            NewsDto finalEntity = NewsDto.builder()
```
Comment (dhanyesh) [inline newsService/src/main/java/com/kovan/app/service/NewsService.java]: change name entity to dto

**Review by dhanyesh — CHANGES_REQUESTED** [review]

File: mongo-dao/src/main/java/com/kovan/dto/NewsDto.java (hunk 11 above)
Comment (Arundhathy-k) [reply mongo-dao/src/main/java/com/kovan/dto/NewsDto.java:21]: @id annotation removed

File: newsService/src/main/java/com/kovan/app/service/NewsService.java (hunk 12 above)
Comment (Arundhathy-k) [reply newsService/src/main/java/com/kovan/app/service/NewsService.java]: Changed name finalEntity to finalDto

**Comment by Arundhathy-k** [comment]
Screenshot (143)

File: newsService/src/main/java/com/kovan/app/controller/NewsServiceController.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+@RestController
+@RequestMapping("/api/news")
+public class NewsServiceController {
+
+    @Autowired
+    private NewsService newsService;
+
+
+    @GetMapping("/top-headlines")
+    public ResponseEntity<NewsDto> fetchAndSaveTopHeadlines() {
```
Comment (Arundhathy-k) [inline newsService/src/main/java/com/kovan/app/controller/NewsServiceController.java:19]: asd

**Pull request merged**